consensus/hotstuff/committees/cluster_committee.go

```diff
@@ -62,8 +62,7 @@ func NewClusterCommittee(
 		clusterMemberFilter: filter.And[flow.Identity](
 			// adapt the identity filter to the identity skeleton filter
 			filter.Adapt(initialClusterMembersSelector),
-			filter.Not(filter.Ejected),
-			filter.HasWeight(true),
+			filter.IsValidCurrentEpochParticipant,
```

Comment:
## ⚠️ 

The previous code would filter out all nodes that had zero weight, which includes nodes with `InitialWeight=0` that could be included in the epoch setup event.

It is a degenerate case, but we **need a clear convention for it and consistently follow that convention**.

The current status quo was that the code would tolerate the Epoch setup event specifying nodes with `InitialWeight=0`.
* The consensus leader selection would never select such nodes, even if they are part of the weighted input set.
* The previous implementation of  ClusterCommittee would also just not consider such nodes valid _active_ participants ... because in a sense they cannot contribute to advancing the protocol. 

I am inclined to retain that convention.
```suggestion
			filter.IsValidCurrentEpochParticipant,
			filter.HasInitialWeight[flow.Identity](true),
```

Furthermore, note that we also need to apply this convention to [`constructInitialClusterIdentities()`](https://github.com/onflow/flow-go/blob/7391fc32b3889e93a4f8bef91f036af5340e1f1c/consensus/hotstuff/committees/cluster_committee.go#L194). The old code also had inconsistent behaviour in this detail.

Reply:
I am inclined to consider an `EpochSetup` event with participants which has `InitialWeight == 0` as an invalid one. A node with zero `InitialWeight` is a node without stake which shouldn't part of the internal network. Non-staked nodes should be part of the external network where they can join without taking part in the staking auction. Such nodes can become part of the internal network and there is zero risk for them to be slashed for their misbehavior, additionally there is a large surface for attack vectors where we need to not forget to filter out nodes with `InitialWeight == 0` to mitigate them. If you still think that it's better to have them as part of identity table then we should very clearly define what such nodes are allowed to do, even in such case I would be inclined to keep the existing behavior and filter out them from both consensus algorithms and leader selection([it's true for consensus committee](https://github.com/onflow/flow-go/blob/yurii/4649-remove-dynamic-weight/consensus/hotstuff/committees/consensus_committee.go#L58C54-L58C80)).

Reply:
Considering `EpochSetup` event with zero-weight participants as invalid is fine by me. Just forbidding zero-weight nodes would probably be the easiest from an engineering perspective. In case we decide to go down that route, we should include corresponding checks in the epoch state machine. 

However, I'd like to emphasize that there will probably not be an easy way to change that convention later:
* The existing software covers the zero-weight scenario. There are a few remaining uncovered edge cases, but the bulk of the logic is already at the correct places, because we considered the dynamic weight reaching zero, which includes the case of the weight being initialized to zero due to `InitialWeight == 0`. 
* At the moment, your PR still checks for positive initial weight in many places. These weight checks would all be obsolete, if we rule out zero weighted nodes entirely, right? 
* So it will be much much more work later to put zero-weight checks back in later, compared to keeping them and filling in the few remaining gaps. 

I am wondering if currently ANs have zero weight, but I am not sure. I think this might be a topic we have to bring up in TPM.

Reply:
My suggestion for the scope of this PR would be to keep the option of zero-weight participants open and try to clarify if we can simplify the approach. Same as you, I feel filtering zero-weight nodes out from both consensus algorithms and leader selection would be the most secure and conceptually consistent approach.

Reply:
I was very tempted by the idea of putting a single check in https://github.com/onflow/flow-go/blob/yurii/4649-remove-dynamic-weight/state/protocol/validity.go#L46 which solves many of our problems and making code simpler. But you are completely right that it's easy to remove but hard to put it back if it's needed again. I wasn't planning on removing checks at this point but I agree that this change is out of the scope of this PR. 
I will fix filtering where applicable so we just filter out nodes with 0 weight for both cluster and main consensus(we already do). 
Thanks for the details!

Reply:
https://github.com/onflow/flow-go/pull/5039/commits/43a21135de8db33541658d5ceb7f95b3295b8382

Reply:
What about the cluster assignment? Do we check weights when building it? Including 0 weight nodes into clustering means that some clusters might have less valid nodes for consensus participation than others.

Reply:
> What about the cluster assignment? Do we check weights when building it?

No, we just [round-robin them irrespective of weights](https://github.com/onflow/flow-core-contracts/blob/master/contracts/epochs/FlowEpoch.cdc#L804-L844) in the smart contract, because it assigns all nodes the same weight.

Reply:
From my perspective, it is fine to have clusters of different sizes (and up to the epoch smart contract to decide how to partition the collectors into clusters). Each cluster is a consensus committee. So I would apply the same reasoning as for the main consensus committee: zero-weight nodes are tolerated, but the `ClusterCommittee` filters them out because they can't contribute meaningful signatures to anything.

Reply:
brought this topic up at yesterday at our Flow Product Meeting. There wasn't any clear opinion. I think we need to bring this question up at TPM before we  decide for good that zero-weight nodes are not allowed. 
The next TPM is in 2 weeks ... by that time we are probably through with this PR. Therefore, I think it might be acceptable to keep the decision open of whether or not we allow zero-weight nodes
* we continue to support it in all places that we are aware of
* in practise, zero-weight nodes might not be happening. Therefore it might be an acceptable risk to: